src/MobileRedirection.php

```diff
@@ -0,0 +1,530 @@
+	 */
+	public function render_setting_field() {
+		?>
+		<p>
+			<label for="mobile_redirect">
+				<input id="mobile_redirect" type="checkbox" name="<?php echo esc_attr( AMP_Options_Manager::OPTION_NAME . '[mobile_redirect]' ); ?>" <?php checked( AMP_Options_Manager::get_option( Option::MOBILE_REDIRECT ) ); ?>>
+				<?php esc_html_e( 'Redirect mobile visitors to the AMP version of a page.', 'amp' ); ?>
+			</label>
+		</p>
+		<script>
+			( function( $ ) {
+				const standardModeInput = $( '#theme_support_standard' );
+				const templateModeInputs = $( 'input[type=radio][name="amp-options[theme_support]"]' );
+				const mobileRedirectSetting = $( 'tr.amp-mobile-redirect' );
+
+				function toggleMobileRedirectSetting( ) {
+					mobileRedirectSetting.toggleClass( 'hidden', standardModeInput.prop( 'checked' ) )
+				}
+
+				templateModeInputs.on( 'change', toggleMobileRedirectSetting );
+				toggleMobileRedirectSetting();
+			} )( jQuery )
+		</script>
+		<?php
+	}
+
+	/**
+	 * Sanitize options.
+	 *
+	 * @param array $options     Existing options with already-sanitized values for updating.
+	 * @param array $new_options Unsanitized options being submitted for updating.
+	 *
+	 * @return array Sanitized options.
+	 */
+	public function sanitize_options( $options, $new_options ) {
+		if ( isset( $new_options[ Option::MOBILE_REDIRECT ] ) && 'on' === $new_options[ Option::MOBILE_REDIRECT ] ) {
+			$options[ Option::MOBILE_REDIRECT ] = true;
+		} else {
+			$options[ Option::MOBILE_REDIRECT ] = false;
+		}
+		return $options;
+	}
+
+	/**
+	 * Get the AMP version of the current URL.
+	 *
+	 * @return string AMP URL.
+	 */
+	private function get_current_amp_url() {
+		$url = add_query_arg( amp_get_slug(), '1', amp_get_current_url() );
+		$url = remove_query_arg( self::NO_AMP_QUERY_VAR, $url );
+		return $url;
+	}
+
+	/**
+	 * Add redirection logic if available for request.
+	 */
+	public function redirect() {
+		if ( ! $this->is_available_for_request() ) {
+			return;
+		}
+
+		$js = $this->is_using_client_side_redirection();
+
+		// If using server-side redirection, make sure that caches vary by user agent.
+		if ( ! $js && ! headers_sent() ) {
+			header( 'Vary: User-Agent', false );
+		}
+
+		// Print the mobile switcher styles.
+		add_action( 'wp_head', [ $this, 'add_mobile_version_switcher_styles' ] );
+		add_action( 'amp_post_template_head', [ $this, 'add_mobile_version_switcher_styles' ] );
+
+		if ( ! is_amp_endpoint() ) {
+			add_action( 'wp_head', [ $this, 'add_mobile_alternative_link' ] );
+			if ( $js ) {
+				// Add mobile redirection script.
+				add_action( 'wp_head', [ $this, 'add_mobile_redirect_script' ], ~PHP_INT_MAX );
+			} elseif ( ! $this->is_redirection_disabled_via_cookie() ) {
+				if ( $this->is_redirection_disabled_via_query_param() ) {
+					// Persist disabling mobile redirection for the session if redirection is disabled for the current request.
+					$this->set_cookie_to_disable_redirection();
```

Comment:
When accessing the AMP version, should the reverse of `set_cookie_to_disable_redirection` be done when server-side redirection is being performed? Should the cookie get cleared? Currently it is getting cleared via the logic in the `onclick` attribute on the version switcher link.

Reply:
Which would go in the `else` statement below, naturally.

Reply:
Done in 7b46bae117a1c5432c972d4d0fa126f8bc4025af